File: src/csharp/examples/generic_scan/Sum.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Alea.CUDA;
using Alea.CUDA.IL;

namespace Tutorial.Cs.examples.generic_scan
{
    public class Sum<T> : ILGPUModule
    {
        public Func<int, T, T, T> MultiScan(int numWarps, int logNumWarps, Func<T,T,T> sum)
        {
            var warpStride = Const.WARP_SIZE + Const.WARP_SIZE/2 + 1;
            return
                (tid, x, totalRef) =>
                {
                    var warp = tid/Const.WARP_SIZE;
                    var lane = tid & (Const.WARP_SIZE - 1);

                    // Allocate shared memory.
                    var shared =
                        Intrinsic.__ptr_volatile(
                            Intrinsic.__array_to_ptr(
                                __shared__.Array<T>(numWarps * warpStride)));

                    var totalsShared =
                        Intrinsic.__ptr_volatile(
                            Intrinsic.__array_to_ptr(
                                __shared__.Array<T>(2*numWarps)));

                    var warpShared = (shared + warp*warpStride).Volatile();
                    var s = (warpShared + lane + Const.WARP_SIZE/2).Volatile();
                    warpShared[lane] = default(T);
                    s[0] = x;

                    // Run inclusive scan on each warp's data.
                    var scan = Intrinsic.__unbox(x);
                    for (var i = 0; i < Const.LOG_WARP_SIZE; i++)
                    {
                        var offset = 1 << i;
                        scan = sum(scan, s[-offset]);
                        if (i < Const.LOG_WARP_SIZE - 1)
                            s[0] = scan;
                    }

                    if (lane == Const.WARP_SIZE - 1)
                        totalsShared[numWarps + warp] = scan;

                    // Synchronize to make all the totals available to the reduction code.
                    Intrinsic.__syncthreads();

                    if (tid < numWarps)
                    {
                        // Grab the block total for the tid'th block. This is the last element
                        // in the block's scanned sequence.  this operation avoids bank conflicts.
                        var total = totalsShared[numWarps + tid];
                        totalsShared[tid] = default(T);
                        var ss = (totalsShared + numWarps + tid).Volatile();

                        var totalsScan = Intrinsic.__unbox(total);
                        for (var i = 0; i < logNumWarps; i++)
                        {
                            var offset = 1 << i;
                            totalsScan = sum(totalsScan, ss[-offset]);
                            ss[0] = totalsScan;
                        }

                        // Store totalsScan shifted by one to the right for an exclusive scan.
                        totalsShared[tid + 1] = totalsScan;
                    }

                    // Synchronize to make the block scan available to all warps
                    Intrinsic.__syncthreads();
                    
                    // The total is the last element.
                    totalRef = totalsShared[2*numWarps - 1];

                    // Add the block scan to the inclusive scan for the block.
                    return sum(scan, totalsShared[warp]);
                };
        } 

        public Sum(GPUModuleTarget target) : base(target)
        {
        }
    }
}

File: src/fsharp/examples/generic_reduce/Sum.fs
(*** hide ***)
module Tutorial.Fs.examples.genericReduce.Sum

open Microsoft.FSharp.Quotations
open Alea.CUDA
open Alea.CUDA.Utilities
open Plan

(**
Sum multi-reduce function for all warps in the block.  This sum-specialized multi reduce function
is unique to the F# implementation of generic reduce.
*)
let inline multiReduce numWarps logNumWarps =
    let warpStride = WARP_SIZE + WARP_SIZE / 2 + 1
    let sharedSize = numWarps * warpStride

    <@ fun tid (x:'T) ->
        let warp = tid / WARP_SIZE
        let lane = tid &&& (WARP_SIZE - 1)
        let shared = __shared__.Array<'T>(sharedSize) |> __array_to_ptr |> __ptr_volatile 
        let warpShared = shared + warp * warpStride     
        let s = warpShared + (lane + WARP_SIZE / 2)

        warpShared.[lane] <- 0G
        s.[0] <- x

        // Run inclusive scan on each warp's data.
        let mutable sum = x |> __unbox
        for i = 0 to LOG_WARP_SIZE - 1 do
            let offset = 1 <<< i
            sum <- sum + s.[-offset]   
            if i < LOG_WARP_SIZE - 1 then s.[0] <- sum
        
        let totalsShared = __shared__.Array<'T>(2*numWarps) |> __array_to_ptr |> __ptr_volatile 

        if lane = WARP_SIZE - 1 then
            totalsShared.[numWarps + warp] <- sum  

        // Synchronize to make all the totals available to the reduction code.
        __syncthreads()

        if tid < numWarps then
            // Grab the block total for the tid'th block. This is the last element
            // in the block's scanned sequence. This operation avoids bank conflicts.
            let total = totalsShared.[numWarps + tid]
            totalsShared.[tid] <- 0G
            let s = totalsShared + numWarps + tid

            let mutable totalsSum = total |> __unbox
            for i = 0 to logNumWarps - 1 do
                let offset = 1 <<< i
                totalsSum <- totalsSum + s.[-offset]
                s.[0] <- totalsSum

        // Synchronize to make the block scan available to all warps.
        __syncthreads()

        // The total is the last element.
        totalsShared.[2 * numWarps - 1] @>

(**
Sum reduces ranges and store reduced values in array of the range totals.  This sum-specialized 
upsweep kernel is unique to the F# implementation of generic reduce.
*)
let inline reduceUpSweepKernel (plan:Plan) =
    let numThreads = plan.NumThreads
    let numWarps = plan.NumWarps
    let logNumWarps = log2 numWarps
    let multiReduce = multiReduce numWarps logNumWarps

    <@ fun (dValues:deviceptr<'T>) (dRanges:deviceptr<int>) (dRangeTotals:deviceptr<'T>) ->
        let block = blockIdx.x
        let tid = threadIdx.x
        let rangeX = dRanges.[block]
        let rangeY = dRanges.[block + 1]

        // Loop through all elements in the interval, adding up values.
        // There is no need to synchronize until we perform the multireduce.
        let mutable sum = 0G |> __unbox
        let mutable index = rangeX + tid
        while index < rangeY do              
            sum <- sum + dValues.[index] 
            index <- index + numThreads

        // Get the total.
        let total = (%multiReduce) tid sum 

        if tid = 0 then dRangeTotals.[block] <- total @>

(**
Sum reduces range totals to a single total, which is written back to the first element in the 
range totals input array.   This sum-specialized reduce range totals kernel is unique to the F# 
implementation of generic reduce.
*)
let inline reduceRangeTotalsKernel (plan:Plan) =
    let numThreads = plan.NumThreadsReduction
    let numWarps = plan.NumWarpsReduction
    let logNumWarps = log2 numWarps
    let multiReduce = multiReduce numWarps logNumWarps

    <@ fun numRanges (dRangeTotals:deviceptr<'T>) ->
        let tid = threadIdx.x

        let x = if tid < numRanges then dRangeTotals.[tid] else 0G
        let total = (%multiReduce) tid x

        // Have the first thread in the block set the range total.
        if tid = 0 then dRangeTotals.[0] <- total @>

